### PR TITLE
Fix issue creating dimensional constraints

### LINKED ISSUE #5456: [Bug]: Inconsistent snapshots when scanning metadata
### What type of bug is this?

Unexpected error

### What subsystems and features are affected?

Data ingestion

### What happened?

When ingesting data and creating chunks under high concurrency (e.g., many clients inserting), the following error can occur:

```
023-03-16 12:22:24.322 CET: enordstr [491028] data_node_1 CONTEXT:  PL/pgSQL function _timescaledb_internal.dimension_slice_get_constraint_sql(integer) line 14 at RAISE
        PL/pgSQL function _timescaledb_internal.chunk_constraint_add_table_constraint(_timescaledb_catalog.chunk_constraint) line 16 at assignment
2023-03-16 12:22:24.322 CET: enordstr [491028] data_node_1 LOCATION:  exec_stmt_raise, pl_exec.c:3873
2023-03-16 12:22:24.322 CET: enordstr [491028] data_node_1 STATEMENT:  SELECT * FROM _timescaledb_internal.create_chunk($1, $2, $3, $4, $5)

```

The reason this happens is because the chunks constraints are added using an upcall to a PL/pgSQL function, which uses a different snapshot (Transaction snapshot) than what is used internally for metadata.  The result is that the scans done in PL/pgSQL code won't see the same information. In this case, a chunk dimension slice exists, but the PL/pgSQL code can't see it.

One solution is to replace this PL/pgSQL code with a proper C function that can scan metadata with a snapshot consistent with the other metadata scans.

### TimescaleDB version affected

2.11.0-dev

### PostgreSQL version used

14.5

### What operating system did you use?

Ubuntu 22.10

### What installation method did you use?

Source

### What platform did you run on?

Other

### Relevant log output and stack trace

_No response_

### How can we reproduce the bug?

```bash
The test I ran was: ingest the NYC tax data set using timescaledb-parallel-copy, 10 workers, into a multi-node cluster with 3 DNs on same local machine and PG instance. Run repeatedly and the problem will occur sooner or later.
```

### PR DESCRIPTION
**Pull request by erimatnor**
During chunk creation, the chunk's dimensional CHECK constraints are created via an "upcall" to PL/pgSQL code. However, creating dimensional constraints in PL/pgSQL code sometimes fails, especially during high-concurrency inserts, because PL/pgSQL code scans metadata using a snapshot that might not see the same metadata as the C code. As a result, chunk creation sometimes fail during constraint creation.

To fix this issue, implement dimensional CHECK-constraint creation in C code. Other constraints (FK, PK, etc.) are still created via an upcall, but should probably also be rewritten in C. However, since these constraints don't depend on recently updated metadata, this is left to a future change.

Fixes #5456

### CODE REVIEW

**Comment by erimatnor** [comment]
I have not  been able to recreate this bug in an isolation test, because it only happens when running really high-concurrency inserts (e.g., parallel-copy with many workers).

Still, this is valuable just as a refactor, getting rid of some very legacy PL/pgSQL code. We should try to get rid of the rest of this upcall code in the future.

File: src/chunk.c [hunk 1]
```diff
@@ -4756,8 +4751,28 @@ ts_chunk_merge_on_dimension(Chunk *chunk, const Chunk *merge_chunk, int32 dimens
...
+	ts_chunk_constraint_scan_iterator_set_slice_id(&iterator, new_slice->fd.id);
+
+	ts_scanner_foreach(&iterator)
+	{
+		bool isnull;
+		Datum d;
+
+		if (slot_attisnull(ts_scan_iterator_slot(&iterator),
+						   Anum_chunk_constraint_dimension_slice_id))
+			continue;
```
Comment (akuzm) [inline src/chunk.c:4766]: Is this possible if we're scanning by dimension slice id?

**Review by akuzm — APPROVED** [review]
> I have not been able to recreate this bug in an isolation test, because it only happens when running really high-concurrency inserts (e.g., parallel-copy with many workers).

Maybe we should start adding probabilistic tests that just simulate this high-concurrency situation for a couple of minutes. Still better than nothing. Probably can be done with TAP tests.

File: src/chunk.c (hunk 1 above)
Comment (erimatnor) [reply src/chunk.c:4766]: No, I guess it is not so I will remove this check. I believe I based this on code from a more general scanning function.

**Review by nikkhils — APPROVED** [review]
LGTM overall.

File: sql/updates/reverse-dev.sql [hunk 2]
```diff
@@ -12,3 +12,90 @@ DROP TABLE IF EXISTS _timescaledb_catalog.continuous_aggs_watermark;
...
+CREATE OR REPLACE FUNCTION _timescaledb_internal.dimension_is_finite(
+    val      BIGINT
+)
+    RETURNS BOOLEAN LANGUAGE SQL IMMUTABLE PARALLEL SAFE AS
+$BODY$
+    --end values of bigint reserved for infinite
+    SELECT val > (-9223372036854775808)::bigint AND val < 9223372036854775807::bigint
+$BODY$ SET search_path TO pg_catalog, pg_temp;
+
+CREATE OR REPLACE FUNCTION _timescaledb_internal.dimension_slice_get_constraint_sql(
```
Comment (pmwkaa) [inline sql/updates/reverse-dev.sql:25]: I am curious why are we still using SQL functions internally? I am thinking about performance for those calls if they are on a hot path

**Review by pmwkaa — APPROVED** [review]

File: sql/updates/reverse-dev.sql (hunk 2 above)
Comment (erimatnor) [reply sql/updates/reverse-dev.sql:25]: This is just legacy stuff and I think it is way overdue for a refactor into C. It also turned out to not be a great idea from a visibility/snapshot perspective. 

Don't think it is that big of a performance issue, though, given that the function is only called when creating a new chunk.

File: src/chunk_constraint.c [hunk 3]
```diff
@@ -312,46 +424,98 @@ chunk_constraint_create(const ChunkConstraint *cc, Oid chunk_oid, int32 chunk_id
...
+	{
+		const ChunkConstraint *cc = &ccs->constraints[i];
+
+		if (is_dimension_constraint(cc))
+		{
+			const DimensionSlice *slice = get_slice_with_id(chunk->cube, cc->fd.dimension_slice_id);
+			const Dimension *dim;
+			Constraint *constr;
+
+			dim = ts_hyperspace_get_dimension_by_id(ht->space, slice->fd.dimension_id);
```
Comment (mkindahl) [inline src/chunk_constraint.c:485]: Variable `slice` can in theory be NULL here if there is something wrong with the metadata. You might want to ensure that it is not NULL before starting to de-reference it.

```suggestion
			Ensure(slice, "dimension slice was missing");
			dim = ts_hyperspace_get_dimension_by_id(ht->space, slice->fd.dimension_id);
```

**Pull request merged**